### PR TITLE
15 new rumour clustering

### PR DESCRIPTION
**Pull request by pius-ng3a**
---

Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [ ] I have checked the code runs
- [ ] I have tested the code
- [ ] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [ ] I have explained the feature in this PR or (better) in `output/reports/`
- [ ] I have requested a code review

### CODE REVIEW

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 1]
```diff
@@ -0,0 +1,412 @@
...
+# %% [markdown]
+# # Approach used
+#
+# * Load and clean the dataset
+# * Convert the unlabelled comments into vector of numerical values using a transformer model
+# * Using FAISS library, compute pairwise similarity scores for the comments
+# * Based on the similarity scores, a connectivity graph
+# * Using a community detection algorithm, we generate subgroups/clusters from the graph
+
+# %%
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: These steps should be covered in the requirements file. So should be ok to run without the below commented cells.

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 2]
```diff
@@ -0,0 +1,412 @@
...
+from sentence_transformers import SentenceTransformer
+import time
+import logging
+from sklearn.neighbors import kneighbors_graph
+import networkx as nx
+import cdlib
+from cdlib import algorithms
+
+# %% [markdown]
+# # Approach used
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: This is a nice summary

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+# %%
+project_directory = cci_cameroon.PROJECT_DIR
+
+# %%
+data_df = pd.read_excel(
+    f"{project_directory}/inputs/data/irfc_staff_labelled_data.xlsx"
+)
+
+# %%
+data_df.first_code.unique()
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: So you are just using the results from the first labelling exercise with three codes?

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+# #!pip install stop_words
+# #!pip install googletrans==4.0.0-rc1
+# #!pip install networkx
+# #!pip install cdlib
+
+
+# %%
+project_directory = cci_cameroon.PROJECT_DIR
+
+# %%
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: I know this is an early stage review so the code doesn't need to be refactored yet. It would be useful though to have some more comments throughout explaining the process of whats happening. So its a more clear step by step.

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+data_df = pd.read_excel(
+    f"{project_directory}/inputs/data/irfc_staff_labelled_data.xlsx"
+)
+
+# %%
+data_df.first_code.unique()
+
+# %%
+data_df.comment = [x.lower() for x in data_df.comment]
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: Is this part for the purpose of the community detection model or for the translation step?

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 6]
```diff
@@ -0,0 +1,412 @@
...
+
+# %%
+data["category_id"] = data.first_code.factorize()[0]
+
+# %%
+# checking the distribution of the codes in the data
+data.groupby("first_code").comment.count().plot(kind="bar")
+
+# %%
+data = data.drop_duplicates(["comment"])
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: Have you checked where the duplicate values are? It would be good to know if any are across codes. For example if one comment is in to codes then you might want to consider which one you drop?

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 7]
```diff
@@ -0,0 +1,412 @@
...
+data.reset_index(inplace=True)
+data.shape
+
+# %%
+# Converting the codes into integers
+data["cluster"] = data.code.factorize()[0]
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: I'm not sure why you need this `cluster` column as well as the `category_id` column? Are they not doing the same thing?

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 8]
```diff
@@ -0,0 +1,412 @@
...
+    plt.show()
+
+
+# %%
+indp2 = faiss.index_factory(
+    sentence_embeddings2.shape[1], "Flat", faiss.METRIC_INNER_PRODUCT
+)
+faiss.normalize_L2(sentence_embeddings2)  # to be used for inner product computation
+# setting the number of neighbors to consider for graph connectivity
+neighbors2 = 3
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: Were we discussing with George a method where we use community detection and don't set the number of clusters in advance?

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 9]
```diff
@@ -0,0 +1,412 @@
...
+
+# %% [markdown]
+# # Working with the larger dataset
+
+# %%
+indp = faiss.index_factory(
+    sentence_embeddings.shape[1], "Flat", faiss.METRIC_INNER_PRODUCT
+)
+faiss.normalize_L2(sentence_embeddings)  # to be used for inner product computation
+neighbors = 10
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: Is this the number of comments per cluster? Or the numbers of neighbours to consider for a comment to be grouped?

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 10]
```diff
@@ -0,0 +1,412 @@
...
+
+# %%
+B = kneighbors_graph(
+    distance_matric2, neighbors2, mode="connectivity", include_self=False
+)
+B.toarray()[0]
+
+# %%
+# load the graph
+G2 = nx.from_numpy_matrix(B, create_using=nx.Graph())
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: This is a really nice visual :) 

Do the numbers represent the IDs?

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 11]
```diff
@@ -0,0 +1,412 @@
...
+draw_communities_graph(G2, generate_colors(coms2), coms2)
+
+# %%
+list(to_use2.comment[coms2.communities[0]])  # Observation on hand washing
+
+# %%
+list(to_use2.comment[coms2.communities[2]])  # belief that the disease exists
+
+# %%
+list(to_use2.comment[coms2.communities[1]])  # belief on wearing of masks
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: Where do communities 0 and 1 sit in the visual above? Are they near each other? Interesting that the graph found a subgroup within a code!

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 12]
```diff
@@ -0,0 +1,412 @@
...
+print(G2)
+# visualize the graph
+nx.draw(G2, with_labels=True)
+
+# %%
+# applying a community algorithm to the graph to identify subgroups
+coms2 = algorithms.leiden(G2)
+draw_communities_graph(G2, generate_colors(coms2), coms2)
+
+# %%
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: For the purpose of the demo on Friday it might be worth running these lists through google translate so you can show the comments in english to give people an idea of what comments are in what community group.

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 13]
```diff
@@ -0,0 +1,412 @@
...
+list(to_use2.comment[coms2.communities[0]])  # Observation on hand washing
+
+# %%
+list(to_use2.comment[coms2.communities[2]])  # belief that the disease exists
+
+# %%
+list(to_use2.comment[coms2.communities[1]])  # belief on wearing of masks
+
+# %%
+list(to_use2.comment[coms2.communities[0]])  # Believe masks are not effective
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: I just tried translating these and doesn't seem like they fit the theme 'Believe masks are not effective'? Or at least the ones I checked didn't. They seemed to be talking about lots of different topics. How did you come to the conclusion they are: Believe masks are not effective?

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 14]
```diff
@@ -0,0 +1,412 @@
...
+# visualize the graph
+nx.draw(G2, with_labels=True)
+
+# %%
+# applying a community algorithm to the graph to identify subgroups
+coms2 = algorithms.leiden(G2)
+draw_communities_graph(G2, generate_colors(coms2), coms2)
+
+# %%
+list(to_use2.comment[coms2.communities[0]])  # Observation on hand washing
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: Looking at the below lists (and translating a few from each) I am unsure how they are fitted to the name described in the hashtags?

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 15]
```diff
@@ -0,0 +1,412 @@
...
+# %%
+len(coms.communities)
+
+# %%
+for index in range(len(coms.communities)):
+    print(data.comment[coms.communities[index]].index)
+    print("######################NEW Group####")
+
+# %%
+list(
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: Is this an example of comments from one of the communities found from the community detection method?

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 16]
```diff
@@ -0,0 +1,412 @@
...
+
+# %%
+indp.train(sentence_embeddings)
+indp.add(sentence_embeddings)
+distance_matric, position = indp.search(
+    sentence_embeddings, sentence_embeddings.shape[0]
+)  # create similarity matrix for the data
+
+# %%
+# create a KNN network with n neighbors using the cosine similarity matrix
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: A bit confused of the difference between the 10 neighbours defined above and 5 in the kneighbors_graph?

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 17]
```diff
@@ -0,0 +1,412 @@
...
+    print(data.comment[coms.communities[index]].index)
+    print("######################NEW Group####")
+
+# %%
+list(
+    data.iloc[[158, 161, 162, 163, 164, 165, 166, 167, 168, 171]].comment
+)  # belief about mask wearing by children
+
+# %%
+list(
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: This community / group seems much better defined than the ones from the smaller dataset.

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 18]
```diff
@@ -0,0 +1,412 @@
...
+            80,
+            83,
+            84,
+            338,
+        ]
+    ].comment
+)
+
+# %%
+## Using walktrap community detection algorithm on the data subset
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: Is there a reason for testing on such a small subset? How are the results below? (I translated a few but hard to get a sense). I noticed the results seemed less affective using the previous method on the smaller subset compared to the larger dataset - so maybe you need to test on the larger aswell?

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py [hunk 19]
```diff
@@ -0,0 +1,412 @@
...
+
+# %%
+list(to_use2.comment[comsw2.communities[3]])
+
+# %%
+
+# %% [markdown]
+# ## General observations
+#
+# * The higher the number of neigbors in the graph, the bigger the clusters formed. Need to agree on an average number of comments that should come through before algorithm is run in order to choose a suitable value for the number of neighbors.
```
Comment (izzyStewart) [inline cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: On this point, I think the idea is that we would run the community detection algorthm each time new comments come through but run it on all comments in the pool of unlabelled comments? So, for example, you may have a pool of 500 unlabelled comments and 2 new ones come through, so you would run the algorithm on 502 comments in that case. Maybe a good idea is to test on different size datasets and evaluate the performance and then we could have a step that checks the number of comments in the dataset and assigns k accordingly?

**Review by izzyStewart — COMMENTED** [review]
Nice work. A few things:

1. Have you found a way to assess the quality of the communities detected other than looking at the comments in each? It feels like this is the missing step to it being a workflow where you need:
- Dataset of unlabelled comments read in
- Dataset cleaned and in format for input to algorithm
- Communities detected
- Identify the communities that fit a predefined criteria: are big enough (eg over 10 comments) and have a evaluation score over a pre set threshold
- Store the comments for the communities that are good enough (as defined above) and save the results in a appropriate file format

2. As you are using a few codes that we decided are fairly well grouped (out of workshop 1), can you identify the difference between the codes and the communities detected? Eg in community 1, what is the breakdown of comments in each code group? For example if it was 100% about the masks code that would tell us it has captured the masks code very well.

3. I think it might be worth trying on a larger subset of the full IFRC dataset. Or, once you have a good workflow setup, try bringing in more code groups and see how it handles it. In a way there is one form of evaluation in terms of how well it groups according to the existing codes but another of just how good a quality the communities are - eg there may be cases where new, good quality groups are found that are subsets or combinations of multiple codes.

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py (hunk 16 above)
Comment (pius-ng3a) [reply cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: There is a subset of the data (~27 comments) that is used in one case. For this, we consider 3 or 5 nearest neighbours while for the larger set of about 358 comments we consider 10 nearest neighbours per data point.

File: cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py (hunk 1 above)
Comment (pius-ng3a) [reply cci_cameroon/analysis/clustering_rumours/rumour_clustering_template.py]: I had included the needed libraries in the requirements.txt file. Would remove the commented section.
